presto-native-execution/presto_cpp/main/expression/RowExpressionOptimizer.cpp

```diff
@@ -0,0 +1,698 @@
+              // This is the first case statement that evaluates to true, so
+              // return the expression corresponding to this case.
+              return {
+                  json::array({veloxToPrestoRowExpression(
+                      caseValue, inputWhenArgs[1])}),
+                  false};
+            } else {
+              // If the case has been constant folded to false in the Velox
+              // switch expression, we do not have access to the expression
+              // inputs in Velox anymore. So we return the corresponding
+              // argument from the input switch expression.
+              result.emplace_back(inputArgs[argsIdx]);
+            }
+          } else {
+            // Skip cases that evaluate to false from the output switch
+            // expression's arguments.
+            continue;
+          }
+        } else {
+          whenArgs.emplace_back(getConstantRowExpression(constantExpr));
+        }
+      } else {
+        VELOX_USER_CHECK(!exprInputs[i]->inputs().empty());
+        const auto& matchExpr = exprInputs[i]->inputs().back();
+        whenArgs.emplace_back(
+            veloxToPrestoRowExpression(matchExpr, inputWhenArgs[0]));
+      }
+
+      whenArgs.emplace_back(
+          veloxToPrestoRowExpression(caseValue, inputWhenArgs[1]));
+      result.emplace_back(getWhenSpecialForm(whenArgs, i));
+    }
+  }
+
+  // Else clause.
+  if (numInputs % 2 != 0) {
+    result.push_back(veloxToPrestoRowExpression(
+        exprInputs[numInputs - 1], inputArgs[numArgs - 1]));
+  }
+  return {result, true};
+}
+
+json RowExpressionConverter::getSpecialForm(
+    const exec::ExprPtr& expr,
+    const json& input) {
+  json res;
+  res["@type"] = kSpecial;
+  auto form = expr->name();
+  // Presto requires the field form to be in upper case.
+  std::transform(form.begin(), form.end(), form.begin(), ::toupper);
+  res["form"] = form;
+  auto exprInputs = expr->inputs();
+  res["arguments"] = json::array();
+
+  // Arguments for switch expression include special form expression 'when'
+  // so it is constructed separately. If the switch expression evaluation found
+  // a case that always evaluates to true, the second value in pair switchResult
+  // will be false and the first value in pair will contain the value
+  // corresponding to the simplified case.
+  if (form == kSwitch) {
+    auto switchResult = getSwitchSpecialFormArgs(expr, input);
+    if (switchResult.second) {
+      res["arguments"] = switchResult.first;
+    } else {
+      return switchResult.first.front();
+    }
+  } else {
+    json::array_t inputArguments = input["arguments"];
+    const auto numInputs = exprInputs.size();
+    VELOX_USER_CHECK_LE(numInputs, inputArguments.size());
+    for (auto i = 0; i < numInputs; i++) {
+      res["arguments"].push_back(
+          veloxToPrestoRowExpression(exprInputs[i], inputArguments[i]));
+    }
+  }
+  res["returnType"] = getTypeSignature(expr->type());
```

Comment:
This line can be moved right to the beginning before the conversion of the inputs.